wgpu-native/src/registry.rs

```diff
@@ -1,47 +1,100 @@
+use std::marker::PhantomData;
+use std::os::raw::c_void;
 use std::sync::{Arc, Mutex};
 use std::{borrow, cmp, fmt, ops, ptr};
 
 use hal::backend::FastHashMap;
 use {AdapterHandle, DeviceHandle, InstanceHandle, ShaderModuleHandle};
 
+#[cfg(not(feature = "remote"))]
+pub(crate) type Id = *mut c_void;
+#[cfg(feature = "remote")]
 pub(crate) type Id = u32;
 
-pub(crate) struct Registry<T> {
+pub(crate) trait Registry<T> {
+    fn new() -> Self;
+    fn register(&mut self, handle: T) -> Id;
+    fn get(&self, id: Id) -> Option<&T>;
+    fn get_mut(&mut self, id: Id) -> Option<&mut T>;
+}
+
+#[cfg(not(feature = "remote"))]
+pub(crate) struct LocalRegistry<T> {
+    marker: PhantomData<T>,
+}
+
+#[cfg(not(feature = "remote"))]
+impl<T> Registry<T> for LocalRegistry<T> {
+    fn new() -> Self {
+        LocalRegistry {
+            marker: PhantomData,
+        }
+    }
+
+    fn register(&mut self, handle: T) -> Id {
+        ::std::boxed::Box::into_raw(Box::new(handle)) as *mut _ as *mut c_void
+    }
+
+    fn get(&self, id: Id) -> Option<&T> {
+        unsafe { (id as *const T).as_ref() }
+    }
+
+    fn get_mut(&mut self, id: Id) -> Option<&mut T> {
+        unsafe { (id as *mut T).as_mut() }
+    }
+}
+
+#[cfg(feature = "remote")]
+pub(crate) struct RemoteRegistry<T> {
     next_id: Id,
     tracked: FastHashMap<Id, T>,
 }
 
-impl<T> Registry<T> {
+#[cfg(feature = "remote")]
+impl<T> Registry<T> for RemoteRegistry<T> {
     fn new() -> Self {
-        Registry {
+        RemoteRegistry {
             next_id: 0,
             tracked: FastHashMap::default(),
         }
     }
 
-    pub(crate) fn register(&mut self, handle: T) -> Id {
+    fn register(&mut self, handle: T) -> Id {
         let id = self.next_id;
         self.tracked.insert(id, handle);
         self.next_id += 1;
         id
     }
 
-    pub(crate) fn get(&self, id: Id) -> Option<&T> {
+    fn get(&self, id: Id) -> Option<&T> {
         self.tracked.get(&id)
     }
 
-    pub(crate) fn get_mut(&mut self, id: Id) -> Option<&mut T> {
+    fn get_mut(&mut self, id: Id) -> Option<&mut T> {
         self.tracked.get_mut(&id)
     }
 }
 
+#[cfg(not(feature = "remote"))]
+lazy_static! {
+    pub(crate) static ref ADAPTER_REGISTRY: Mutex<LocalRegistry<AdapterHandle>> =
```

Comment:
It would be great to drop this `Mutex` too but it's not obvious how to support both paths without it. We need a no-op mutex or some helper functions to wrap usage of the mutex for the `remote` case.

Reply:
I think we should just move the mutability (i.e. `Mutex`) into the internals of the registry implementation, so that registry itself is Send+Sync and most methods are `&self`